File: ios/LLMHub/Sources/LLMHub/MimoBot/OpusCodec.swift
import Foundation

/// Opus encode/decode at 16 kHz mono, 20 ms frames.
///
/// iOS doesn't have a system Opus API. Options:
///   - YbridOpus (https://github.com/ybrid/opus-swift) — Swift wrapper over libopus,
///     binary-compatible with the Android side.
///   - Roll a thin XCFramework from the upstream libopus source.
///   - CocoaPods `libopus`.
///
/// Starting choice: YbridOpus as a SwiftPM dep (cleanest). Add to Package.swift
/// dependencies when we're ready to wire it in.
///
/// TODO(opus):
///   1. Add `.package(url: "https://github.com/ybrid/opus-swift", from: "0.8.0")`
///      to ios/LLMHub/Package.swift (only when implementing — avoid extra deps
///      until then so the tree keeps building).
///   2. Replace the bodies below with the real encoder/decoder calls.
///   3. Reuse encoder/decoder instances.
final class OpusEncoderWrap {
    let sampleRateHz: Int
    let channels: Int
    let bitrateBps: Int
    let frameSamples: Int

    init(sampleRateHz: Int = 16_000, channels: Int = 1, bitrateBps: Int = 24_000, frameSamples: Int = 320) {
        self.sampleRateHz = sampleRateHz
        self.channels = channels
        self.bitrateBps = bitrateBps
        self.frameSamples = frameSamples
    }

    func encode(pcm: [Int16]) -> Data {
        precondition(pcm.count == frameSamples, "expected \(frameSamples) samples, got \(pcm.count)")
        fatalError("TODO(opus): encode via libopus")
    }
}

final class OpusDecoderWrap {
    let sampleRateHz: Int
    let channels: Int
    let frameSamples: Int

    init(sampleRateHz: Int = 16_000, channels: Int = 1, frameSamples: Int = 320) {
        self.sampleRateHz = sampleRateHz
        self.channels = channels
        self.frameSamples = frameSamples
    }

    func decode(opusFrame: Data) -> [Int16] {
        fatalError("TODO(opus): decode via libopus")
    }

    func decodePLC() -> [Int16] {
        fatalError("TODO(opus): PLC frame via libopus (data=nil)")
    }
}

File: android/app/src/main/java/com/llmhub/llmhub/mimobot/audio/OpusCodec.kt
package com.llmhub.llmhub.mimobot.audio

/**
 * Opus encode/decode for 16 kHz mono, 20 ms frames.
 *
 * We want a JVM Opus binding here. Two reasonable choices:
 *   - concentus (pure-Java, easy to ship, ~3x slower than native — fine at 24 kbps).
 *       implementation("org.concentus:Concentus:0.3.1") or similar artifact.
 *   - an NDK build of libopus exposed via JNI (lower CPU, more work).
 *
 * We'll start with Concentus and switch later if phones struggle. Nothing above
 * this file depends on the choice.
 *
 * TODO(opus):
 *   1. Add the Concentus dependency to android/app/build.gradle.kts:
 *        implementation("org.concentus:Concentus:0.3.1")
 *   2. Replace the bodies below with OpusEncoder / OpusDecoder calls.
 *   3. Reuse Encoder/Decoder instances — they're stateful (FEC, predictor state).
 */
class OpusEncoder(
    val sampleRateHz: Int = 16_000,
    val channels: Int = 1,
    val bitrateBps: Int = 24_000,
    val frameSamples: Int = 320,
) {
    fun encode(pcm: ShortArray): ByteArray {
        require(pcm.size == frameSamples) { "expected $frameSamples samples, got ${pcm.size}" }
        TODO("Concentus OpusEncoder.encode(pcm, 0, frameSamples, outBuf, 0, outBuf.size)")
    }

    fun close() { /* no-op until we wire the encoder */ }
}

class OpusDecoder(
    val sampleRateHz: Int = 16_000,
    val channels: Int = 1,
    val frameSamples: Int = 320,
) {
    fun decode(opusFrame: ByteArray): ShortArray {
        TODO("Concentus OpusDecoder.decode(opusFrame, 0, opusFrame.size, out, 0, frameSamples, false)")
    }

    /** Produce PCM for a lost frame using Opus PLC. */
    fun decodePlc(): ShortArray {
        TODO("Concentus OpusDecoder.decode(null, ..., decodeFec=false)")
    }

    fun close() { /* no-op */ }
}
